### PR TITLE
zips, products and takes with infinite components

### PR DESCRIPTION
**Pull request by mschauer**
This is #13276 using the iterator traits of #15146 (now hopefully rebased).

### CODE REVIEW

**Comment by JeffBezanson** [comment]
:+1: 

Could we have similar definitions for `take` and `drop`?

**Comment by mschauer** [comment]
I added similar definitions for take and drop.

**Comment by mschauer** [comment]
Btw, is there a conceivable situation case where ``Rest`` does not inherit the size trait of its iterator?

File: base/iterator.jl [hunk 1]
```diff
@@ -2,6 +2,16 @@
...
 isempty(itr) = done(itr, start(itr))
 
+_min_length(a, b, ::IsInfinite, ::IsInfinite) = min(length(a),length(b)) # inherit behaviour, error
+_min_length(a, b, A, ::IsInfinite) = length(a)
+_min_length(a, b, ::IsInfinite, B) = length(b)
+_min_length(a, b, A, B) = min(length(a),length(b))
+
+_diff_length(a, b, A, ::IsInfinite) = 0
+_diff_length(a, b, ::IsInfinite, ::IsInfinite) = 0
+_diff_length(a, b, ::IsInfinite, B) = length(b) # inherit behaviour, error
```
Comment (JeffBezanson) [inline base/iterator.jl]: Should this call `length(a)` instead?

Comment (mschauer) [inline base/iterator.jl]: Yes!

**Comment by JeffBezanson** [comment]
The problem with `Rest` is that there's no standard way to determine the length of an iterable starting at a given state, so for now it has to be unknown most of the time.

**Pull request merged**